[WIP] QUARKS-96: postJson in HttpStreams

saurabhjinturkar opened the pull request.

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -46,6 +47,12 @@ Licensed to the Apache Software Foundation (ASF) under one
                 t -> HttpGet.METHOD_NAME, uri, HttpResponders.json());
     }
     
+    public static TStream<JsonObject> postJson(TStream<JsonObject> stream, 
```
please add sufficient javadoc.  extra credit: see https://issues.apache.org/jira/browse/QUARKS-97  :-)

saurabhjinturkar commented:
Hi @dlaboss,

I have added functionality to support postJson() method. I added a method to add body to HTTP requests. Could you please review the code? I will fix review comments and failing build together.

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +40,88 @@ Licensed to the Apache Software Foundation (ASF) under one
+     *     request1.addProperty("a", "abc");
+     *     request1.addProperty("b", "42");
+     *     TStream<JsonObject> rc = HttpStreams.getJson(
+     *             topology.collection(Arrays.asList(request1)),
```
nit: move the construction of the input stream outside of the getJson() invocation?  It's not something that will ever occur in practice.  Ditto the postJson() example below.

saurabhjinturkar replied:
Could you please elaborate on this one? I am not clear on this.

dlaboss replied:
e.g.,
```
TStream<JsonObject> requests = topology.collection(Arrays.asList(request1));
TStream<JsonObject> rc = HttpStreams.getJson(requests, ...)
```

saurabhjinturkar replied:
Oh okay.. i was thinking something complex and weird.

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +40,88 @@ Licensed to the Apache Software Foundation (ASF) under one
+     *     TStream<JsonObject> rc = HttpStreams.getJson(
+     *             topology.collection(Arrays.asList(request1)),
+     *             HttpClients::noAuthentication,
+     *             t -> url + "a=" + t.get("a").getAsString() + "&b="
```
I think it may be clearer for the user if a declaration of the `url` variable is added.

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +40,88 @@ Licensed to the Apache Software Foundation (ASF) under one
+     * <pre>
+     * {@code
+     *     JsonObject request = new JsonObject();
+     *     request.addProperty("a", "abc");
```
Given the getJson() example I was sort of expecting that the "a" and "b" json properties would be used in forming the url.  Maybe tweak this example to rename the properties or to use them the same way?  Same `url` var comment as above.

saurabhjinturkar replied:
Are you suggesting to change the variable name of JsonObject?

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +40,88 @@ Licensed to the Apache Software Foundation (ASF) under one
+    public static TStream<JsonObject> postJson(TStream<JsonObject> stream,
+            Supplier<CloseableHttpClient> clientCreator,
+            Function<JsonObject, String> uri,
+            Function<JsonObject, HttpEntity> body) {
```
It strikes me that the ``body`` function should instead be ``UnaryOperator<JsonObject>`` (a fn that yields a JsonObject (body) from the input JsonObject tuple).  I think that's more in line with "keeping the simple simple" for applications exchanging Json-based data.  I agree that ``HttpEntity`` is appropriate for ``requestsWithBody()``.  Does that make sense?

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/runtime/HttpRequester.java:
```diff
@@ -69,16 +84,35 @@ public R apply(T t) {
+            request = new HttpPost(uri);
+            break;
+        case HttpPut.METHOD_NAME:
+            request = new HttpPut(uri);
```
Adding Put seems like a good idea.  Of course that then begs the question: should there also be {put,delete}Json()? (and tests)  :-)

dlaboss commented on connectors/http/src/test/java/quarks/test/connectors/http/HttpTest.java:
```diff
@@ -170,4 +171,42 @@ public void testJsonGet() throws Exception {
         
         assertTrue(endCondition.getResult().toString(), endCondition.valid());
     }
+    
```
Pls add a testPost() too (e.g., there's a testGet() and testJsonGet()).

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +44,110 @@ Licensed to the Apache Software Foundation (ASF) under one
+            t -> HttpGet.METHOD_NAME, uri, HttpResponders.json());
+    }
+    
+    public static TStream<JsonObject> deleteJson(TStream<JsonObject> stream,
```
javadoc for {delete,put}Json()?

saurabhjinturkar commented:
Hi @dlaboss,
Regarding build failure, it seems there is some issue with SSL and certs. I checked  logs for failed build of #116. I found that its the same reason as for #106. Could you please guide on this?

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +44,191 @@ Licensed to the Apache Software Foundation (ASF) under one
+     * 
+     *     TStream<JsonObject> stream = topology.collection(Arrays.asList(body));
+     *     TStream<JsonObject> rc = HttpStreams.postJson(stream,
+     *             HttpClients::noAuthentication, t -> url, t -> body);
```
I don't think you really want/mean `t -> body`.  Presumably the sample is trying to demonstrate posting each tuple from a stream as the body of a POST to the specified url.  To do that you need to specify 't -> t'  (or `Functions.identity()`), right?

Also the description of `@param body` needs work.  The function isn't returning an HttpEntity anymore.  It's a function that for each JsonObject tuple from stream, returns a JsonObject whose JSON will be the body of the POST request.  Right?

dlaboss commented on connectors/http/src/main/java/quarks/connectors/http/HttpStreams.java:
```diff
@@ -38,12 +44,191 @@ Licensed to the Apache Software Foundation (ASF) under one
+     * 
+     *     TStream<JsonObject> stream = topology.collection(Arrays.asList(body));
+     *     TStream<JsonObject> rc = HttpStreams.putJson(stream,
+     *             HttpClients::noAuthentication, t -> url, t -> body);
```
same comment as above for `t -> body` and `@param body` description

dlaboss commented on connectors/http/src/test/java/quarks/test/connectors/http/HttpTest.java:
```diff
@@ -80,9 +154,9 @@ public void testStatusCode() throws Exception {
         
         String url = "http://httpbin.org/status/";
         
+        TStream<Integer> stream = topology.collection(Arrays.asList(200, 404, 202));
```
Just to clarify, when I mentioned pulling the declaration of the `stream` outside of the request() invocation, I was only talking about for the samples in the javadoc.  It's OK that test code had it that way.  But it's fine to leave it this way now.  Sorry for the confusion.

dlaboss commented on connectors/http/src/test/java/quarks/test/connectors/http/HttpTest.java:
```diff
@@ -170,4 +288,80 @@ public void testJsonGet() throws Exception {
+        TStream<JsonObject> stream = topology.collection(Arrays.asList(body));
+        TStream<JsonObject> rc = HttpStreams.postJson(
+                stream, HttpClients::noAuthentication, t -> url,
+                t -> body);
```
should be `t -> t`

dlaboss commented on connectors/http/src/test/java/quarks/test/connectors/http/HttpTest.java:
```diff
@@ -170,4 +288,80 @@ public void testJsonGet() throws Exception {
+        TStream<JsonObject> stream = topology.collection(Arrays.asList(body));
+        TStream<JsonObject> rc = HttpStreams.putJson(
+                stream, HttpClients::noAuthentication, t -> url,
+                t -> body);
```
should be `t -> t`

dlaboss commented:
sorry for the delay... 
I've created pr #119 to remove these tests that aren't robust in the travis/ci environment.

saurabhjinturkar commented:
Hi @dlaboss, It's final exam week for me. Sorry for the delay in fixing review comments.

dlaboss commented:
no worries. hope things go well for you!

dlaboss commented:
Thanks for your contribution!

saurabhjinturkar commented:
Thank you for your guidance! :)
